File: python/numbers/fibonacci.py
"""Fibonacci sequence — generate to a count or up to a value."""

from typing import Iterator


def fibonacci(n: int) -> list[int]:
    """Return the first n Fibonacci numbers."""
    if n < 0:
        raise ValueError("n must be non-negative")
    if n == 0:
        return []
    if n == 1:
        return [0]

    seq = [0, 1]
    for _ in range(n - 2):
        seq.append(seq[-1] + seq[-2])
    return seq


def fibonacci_up_to(limit: int) -> list[int]:
    """Return all Fibonacci numbers up to and including limit."""
    if limit < 0:
        return []
    seq = [0]
    if limit == 0:
        return seq
    a, b = 0, 1
    while b <= limit:
        seq.append(b)
        a, b = b, a + b
    return seq


def fibonacci_gen() -> Iterator[int]:
    """Infinite Fibonacci generator."""
    a, b = 0, 1
    while True:
        yield a
        a, b = b, a + b


def fib(n: int) -> int:
    """Return the nth Fibonacci number (0-indexed)."""
    if n < 0:
        raise ValueError("n must be non-negative")
    a, b = 0, 1
    for _ in range(n):
        a, b = b, a + b
    return a

File: rust/numbers/fibonacci.rs
//! Fibonacci sequence — generate to a count or up to a value.

/// Return the first `n` Fibonacci numbers.
pub fn fibonacci(n: usize) -> Vec<u128> {
    if n == 0 {
        return vec![];
    }
    if n == 1 {
        return vec![0];
    }

    let mut seq = Vec::with_capacity(n);
    seq.push(0);
    seq.push(1);
    for i in 2..n {
        seq.push(seq[i - 1] + seq[i - 2]);
    }
    seq
}

/// Return all Fibonacci numbers up to and including `limit`.
pub fn fibonacci_up_to(limit: u128) -> Vec<u128> {
    let mut seq = vec![0];
    let (mut a, mut b) = (0u128, 1u128);
    while b <= limit {
        seq.push(b);
        let next = a + b;
        a = b;
        b = next;
    }
    seq
}

/// Return the nth Fibonacci number (0-indexed).
pub fn fib(n: usize) -> u128 {
    let (mut a, mut b) = (0u128, 1u128);
    for _ in 0..n {
        let next = a + b;
        a = b;
        b = next;
    }
    a
}

/// Infinite Fibonacci iterator.
pub struct FibIter {
    a: u128,
    b: u128,
}

impl FibIter {
    pub fn new() -> Self {
        Self { a: 0, b: 1 }
    }
}

impl Iterator for FibIter {
    type Item = u128;

    fn next(&mut self) -> Option<Self::Item> {
        let val = self.a;
        let next = self.a + self.b;
        self.a = self.b;
        self.b = next;
        Some(val)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_fibonacci_10() {
        assert_eq!(fibonacci(10), vec![0, 1, 1, 2, 3, 5, 8, 13, 21, 34]);
    }

    #[test]
    fn test_fib_nth() {
        assert_eq!(fib(0), 0);
        assert_eq!(fib(1), 1);
        assert_eq!(fib(10), 55);
    }

    #[test]
    fn test_up_to() {
        assert_eq!(fibonacci_up_to(10), vec![0, 1, 1, 2, 3, 5, 8]);
    }
}
